launcher/resources/git-shim.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/python
+import socket, os, os.path, sys
+
+PORT=20280
+BUF_SIZE=4096
+
+def receive_all(sock):
+    """Receive all data coming through the socket and return it as bytes."""
+
+    chunk = s.recv(BUF_SIZE)
+    reply = ""
+
+    while len(chunk) != 0:
+        reply += chunk.decode('utf-8')
```

Comment:
I think this `+=` ends up being quadratic as well; maybe we can have an `reply = []` and append chunks to it which we combine via `"".join` before returning? That would bring it down to O(n)

Reply:
I'd like to understand this, are you saying that string concatenation in Python is quadratic? And how is `join` better? I'm only trying to understand how Python works.

Reply:
Anyway, this code has a different problem so it needs to be rewritten (utf-8 characters may not end neatly at chunk boundaries, so decoding has to happen at the end).

Reply:
string concatenation is linear, repeated concatenation is quadratic. OTOH appending to a mutable list is constant time, and "".join is linear, making the total time linear instead of quadratic

Reply:
Ah so strings are immutable! I expected them to be lists, which are mutable in Python. Good to know.